system/lib/fetch/wasmfs.cpp

```diff
@@ -0,0 +1,55 @@
+// // Copyright 2021 The Emscripten Authors.  All rights reserved.
+// // Emscripten is available under two separate licenses, the MIT license and the
+// // University of Illinois/NCSA Open Source License.  Both these licenses can be
+// // found in the LICENSE file.
+
+#include <stdlib.h>
+#define __NEED_struct_iovec
+#include <emscripten/emscripten.h>
+#include <wasi/api.h>
+
+extern "C" {
+
+  __wasi_errno_t __wasi_fd_write( __wasi_fd_t fd, const __wasi_ciovec_t *iovs, size_t iovs_len, __wasi_size_t *nwritten) {
+    if (fd == 1 || fd == 2) {
+      MAIN_THREAD_EM_ASM({
+        var buffer = [];
+        var fd = $0;
+        var iovs = $1;
+        var iovs_len = $2;
+        var nwritten = $3;
+        var num = 0;
+        for (var i = 0; i < iovs_len; i++) {
+          var ptr = HEAP32[(((iovs)+(i*8))>>2)];
+          var len = HEAP32[(((iovs)+(i*8 + 4))>>2)];
```

Comment:
@kripken, is there a better way to write these memory accesses in an `EM_ASM`?

Stepping back, it would much easier to read if all this code that reads and writes memory were just normal C++ rather than JS. There shouldn't be any need to call out to JS for anything except the call to `out` or `err`, and `buffer` can probably be an ArrayBuffer initialized directly from a slice of the Wasm memory rather than built up one character at a time.

Reply:
This is basically the best way from JS, I think. (Though the extra parens can be removed, as I commented after you posted.)

We could do more work in C++ and leave less for JS. That would add additional copying, though, so I'm not sure it's a good idea.